core/src/main/java/com/linecorp/armeria/server/RoutingPredicate.java

```diff
@@ -0,0 +1,236 @@
+import static com.linecorp.armeria.server.RoutingPredicate.ParsedComparingPredicate.parse;
+import static java.util.Objects.requireNonNull;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.function.Predicate;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.StreamSupport;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.MoreObjects;
+
+import com.linecorp.armeria.common.Flags;
+import com.linecorp.armeria.common.HttpHeaders;
+import com.linecorp.armeria.common.HttpParameters;
+
+/**
+ * A {@link Predicate} to evaluate whether a request can be accepted by a service method.
+ * Currently predicates for {@link HttpHeaders} and HTTP parameters are supported.
+ *
+ * @param <T> the type of the object to be tested
+ */
+final class RoutingPredicate<T> implements Predicate<T> {
+    private static final Logger logger = LoggerFactory.getLogger(RoutingPredicate.class);
+
+    /**
+     * Patterns used to parse a given predicate. The predicate can be one of the following forms:
+     * <ul>
+     *     <li>{@code some-name=some-value} which means that the request must have a
+     *     {@code some-name=some-value} header or parameter</li>
+     *     <li>{@code some-name!=some-value} which means that the request must not have a
+     *     {@code some-name=some-value} header or parameter</li>
+     *     <li>{@code some-name} which means that the request must contain a {@code some-name} header or
+     *     parameter</li>
+     *     <li>{@code !some-name} which means that the request must not contain a {@code some-name} header or
+     *     parameter</li>
+     * </ul>
+     */
+    @VisibleForTesting
+    static final Pattern CONTAIN_PATTERN = Pattern.compile("^\\s*([!]?)([^\\s=><!]+)\\s*$");
+    private static final Pattern COMPARE_PATTERN = Pattern.compile("^\\s*([^\\s!><=]+)\\s*([><!]?=|>|<)(.*)$");
+
+    static List<RoutingPredicate<HttpHeaders>> copyOfHeaderPredicates(Iterable<String> predicates) {
+        return StreamSupport.stream(predicates.spliterator(), false)
+                            .map(RoutingPredicate::ofHeaders).collect(toImmutableList());
+    }
+
+    static List<RoutingPredicate<HttpParameters>> copyOfParamPredicates(Iterable<String> predicates) {
+        return StreamSupport.stream(predicates.spliterator(), false)
+                            .map(RoutingPredicate::ofParams).collect(toImmutableList());
+    }
+
+    static RoutingPredicate<HttpHeaders> ofHeaders(CharSequence headerName, Predicate<String> valuePredicate) {
+        return new RoutingPredicate<>(headerName, headers -> {
+            try {
+                return headers.getAll(headerName).stream().anyMatch(valuePredicate);
+            } catch (Throwable cause) {
+                if (Flags.verboseExceptionSampler().isSampled(cause.getClass())) {
```

Comment:
I'm curious whether this usage is okay or not. The reason why I don't log error always is the abnormal request is not intended by our user. It might be from abusing. Let me know if I need to implement this by other way.